### DISCUSSION TITLE
Integrating via C API

### DISCUSSION THREAD

**Discussion by ethindp** [AccessKit/accesskit#351] [Q&A] [answered]
I'm hoping to get started integrating this into a C++ immediate-mode UI framework (either TGUI or Elements, though TGUI might be the one I go with just because it has more event coverage). However, I'm no expert in accessibility myself, though I'm doing my best to learn, but I find the C example a bit confusing. Either immediate-mode UI is backend-agnostic; TGUI supports GLFW, SDL, and SFML, and elements is entirely agnostic to the backend being used. So what I'm trying to ask is: what is the process for integrating AccessKit into a UI library like this? I understand that AccessKit's implementations aren't perfect, but it would be great if I could figure this out and then just wait for AccessKit to catch up.

As an aside, I know that I could just "hook" the widgets (e.g. create an AccessibleButton class or what not) in a consumer application, but this seems counterintuitive and error-prone, particularly since it would then be up to me to keep up with the development of the libraries if their APIs change in a breaking way. Plus, the README says that documentation is needed, so I thought to ask this to kick off that process so people have at least something to go off of when integrating this into a UI library with the C API.

**Comment by DataTriny** [comment, marked as answer]
Hello @ethindp,

To answer your question about windowing providers: AccessKit can't quite be agnostic as none of the currently available libraries expose an API to properly install an accessibility provider. What is required also depends on the platform. Our example shows how to do this with SDL2, but the process will be very similar for SFML and GLFW:

- Windows and macOS: create the window hidden
- Windows and macOS: retrieve the native window handle: [SDL2](https://wiki.libsdl.org/SDL2/SDL_GetWindowWMInfo), [SFML](https://www.sfml-dev.org/documentation/2.6.1/classsf_1_1WindowBase.php#af9e56181556545bf6e6d7ed969edae21), [GLFW](https://www.glfw.org/docs/3.3/group__native.html)
- macOS: you might need to call `accesskit_macos_add_focus_forwarder_to_window_class`, unfortunately you will have to read the source code of the windowing library to find out if you need this, and if yes, find what is the class name
- Initialize AccessKit:
  - `accesskit_windows_subclassing_adapter_new`
  - `accesskit_macos_subclassing_adapter_for_window`
  - `accesskit_unix_adapter_new`
- Windows and macOS: show the window
- macOS and *nix: notify AccessKit when the window gain or lose focus:
  - `accesskit_macos_subclassing_adapter_update_view_focus_state`
  - `accesskit_unix_adapter_update_window_focus_state`
- *nix: notify AccessKit when the window is moved and/or resized: `accesskit_unix_adapter_set_root_window_bounds`
- Everytime something changes in the UI, push either a full accessibility tree or one which only contains what has changed:
  - `accesskit_windows_subclassing_adapter_update_if_active`
  - `accesskit_macos_subclassing_adapter_update_if_active`
  - `accesskit_unix_adapter_update_if_active`

Please refer to the latest version of `accesskit.h`, available in the [latest accesskit_c pre-built package](https://github.com/AccessKit/accesskit/releases/download/accesskit_c-v0.7.1/accesskit_c-v0.7.1.zip).

Since `accesskit_c` is a thin wrapper on top of the Rust libraries, the documentation for the different platform adapters also apply:
- [`accesskit_windows`](https://docs.rs/accesskit_windows/latest/accesskit_windows/)
- [`accesskit_macos`](https://docs.rs/accesskit_macos/latest/accesskit_macos/)
- [`accesskit_unix`](https://docs.rs/accesskit_unix/latest/accesskit_unix/)

The C function names can help you navigate the Rust documentation: `<PACKAGE>_<STRUCT>_<METHOD>`. The above links will also point you to the documentation of the core types shared by all adapters.

**Comment by ethindp** [comment]
@DataTriny Thank you very much for your response!
